flow/include/flow/Arena.h

```diff
@@ -250,16 +252,140 @@ class Optional : public ComposedIdentifier<T, 4> {
 
 	template <class R>
 	Optional<R> castTo() const {
-		return map<R>([](const T& v) { return (R)v; });
+		return map([](const T& v) { return (R)v; });
 	}
 
-	template <class R>
-	Optional<R> map(std::function<R(T)> f) const& {
-		return present() ? Optional<R>(f(get())) : Optional<R>();
+private:
+	template <class F>
+	using MapRet = std::decay_t<std::invoke_result_t<F, T>>;
+
+	template <class F>
+	using EnableIfNotMemberPointer =
+	    std::enable_if_t<!std::is_member_object_pointer_v<F> && !std::is_member_function_pointer_v<F>>;
+
+public:
+	// If the optional is set, calls the function f on the value and returns the value. Otherwise, returns an empty
+	// optional.
+	template <class F, typename = EnableIfNotMemberPointer<F>>
+	Optional<MapRet<F>> map(F f) const& {
+		return present() ? Optional<MapRet<F>>(f(get())) : Optional<MapRet<F>>();
+	}
+	template <class F, typename = EnableIfNotMemberPointer<F>>
+	Optional<MapRet<F>> map(F f) && {
+		return present() ? Optional<MapRet<F>>(f(std::move(*this).get())) : Optional<MapRet<F>>();
+	}
+
+	// Converts an Optional<T> to an Optional<R> of one of its value's members
+	//
+	// v.map(&T::member) is equivalent to v.map([](T v) { return v.member; })
+	template <class R, class Rp = std::decay_t<R>>
+	std::enable_if_t<std::is_class_v<T>, Optional<Rp>> map(
+	    R std::conditional_t<std::is_class_v<T>, T, Void>::*member) const& {
```

Comment:
I'm confused about why we need the `std::conditional_t` here - don't we kind of know that it's a class from the enable if?

Reply:
If I remember correctly, it is because we want the type of the argument to be `R T::*member`, and even though this function isn't usable if `T` isn't a class, it still tries to process that argument declaration.

By adding in this `conditional_t`, for non-class types it instead tries to process it as `Void::*member`, which is sufficient to fix the compilation. It doesn't really matter much that this doesn't exist since the function can't be used.